File: claude_setup/assembler/patterns_assembler.py
from __future__ import annotations

from pathlib import Path
from typing import List

from claude_setup.domain.pattern_mapping import (
    select_pattern_files,
    select_patterns,
)
from claude_setup.models import ProjectConfig
from claude_setup.template_engine import TemplateEngine

PATTERNS_SKILL_DIR = "patterns"
REFERENCES_DIR = "references"
SKILLS_DIR = "skills"
CONSOLIDATED_FILENAME = "SKILL.md"
SECTION_SEPARATOR = "\n\n---\n\n"


class PatternsAssembler:
    """Assembles pattern knowledge packs from source templates."""

    def __init__(self, src_dir: Path) -> None:
        self._src_dir = src_dir

    def assemble(
        self,
        config: ProjectConfig,
        output_dir: Path,
        engine: TemplateEngine,
    ) -> List[Path]:
        """Orchestrate pattern selection, flush, and consolidation."""
        categories = select_patterns(config)
        if not categories:
            return []
        pattern_files = select_pattern_files(
            self._src_dir, categories,
        )
        if not pattern_files:
            return []
        return self._generate_output(
            pattern_files, output_dir, engine,
        )

    def _generate_output(
        self,
        pattern_files: List[Path],
        output_dir: Path,
        engine: TemplateEngine,
    ) -> List[Path]:
        """Generate individual and consolidated output files."""
        rendered = self._render_contents(pattern_files, engine)
        results: List[Path] = []
        refs_dir = self._build_refs_dir(output_dir)
        results.extend(
            self._flush_patterns(pattern_files, rendered, refs_dir),
        )
        consolidated = self._build_consolidated_path(output_dir)
        results.append(
            self._flush_consolidated(rendered, consolidated),
        )
        return results

    def _render_contents(
        self,
        pattern_files: List[Path],
        engine: TemplateEngine,
    ) -> List[str]:
        """Read and render all pattern files once."""
        rendered: List[str] = []
        for src_file in pattern_files:
            content = src_file.read_text(encoding="utf-8")
            rendered.append(engine.replace_placeholders(content))
        return rendered

    def _build_refs_dir(self, output_dir: Path) -> Path:
        """Build the references directory path."""
        return (
            output_dir / SKILLS_DIR / PATTERNS_SKILL_DIR
            / REFERENCES_DIR
        )

    def _build_consolidated_path(self, output_dir: Path) -> Path:
        """Build the consolidated file path."""
        return (
            output_dir / SKILLS_DIR / PATTERNS_SKILL_DIR
            / CONSOLIDATED_FILENAME
        )

    def _flush_patterns(
        self,
        pattern_files: List[Path],
        rendered: List[str],
        dest_dir: Path,
    ) -> List[Path]:
        """Write pre-rendered pattern files to destination."""
        results: List[Path] = []
        for src_file, content in zip(pattern_files, rendered):
            category = src_file.parent.name
            target_dir = dest_dir / category
            target_dir.mkdir(parents=True, exist_ok=True)
            dest_file = target_dir / src_file.name
            dest_file.write_text(content, encoding="utf-8")
            results.append(dest_file)
        return results

    def _flush_consolidated(
        self,
        rendered: List[str],
        dest_path: Path,
    ) -> Path:
        """Merge pre-rendered content into a single consolidated file."""
        dest_path.parent.mkdir(parents=True, exist_ok=True)
        merged = SECTION_SEPARATOR.join(rendered)
        dest_path.write_text(merged, encoding="utf-8")
        return dest_path

File: src/assembler/patterns-assembler.ts
/**
 * PatternsAssembler — selects and consolidates pattern knowledge packs.
 *
 * Migrated from Python `assembler/patterns_assembler.py` (117 lines).
 * Selection logic lives in {@link ../domain/pattern-mapping.ts}.
 *
 * @module
 */
import * as fs from "node:fs";
import * as path from "node:path";
import type { ProjectConfig } from "../models.js";
import type { TemplateEngine } from "../template-engine.js";
import {
  selectPatterns,
  selectPatternFiles,
} from "../domain/pattern-mapping.js";

const SKILLS_DIR = "skills";
const PATTERNS_SKILL_DIR = "patterns";
const REFERENCES_DIR = "references";
const CONSOLIDATED_FILENAME = "SKILL.md";
const SECTION_SEPARATOR = "\n\n---\n\n";

/** Assembles pattern knowledge packs from source templates. */
export class PatternsAssembler {
  /** Orchestrate pattern selection, rendering, and consolidation. */
  assemble(
    config: ProjectConfig,
    outputDir: string,
    resourcesDir: string,
    engine: TemplateEngine,
  ): string[] {
    const categories = selectPatterns(config);
    if (categories.length === 0) return [];
    const patternFiles = selectPatternFiles(resourcesDir, categories);
    if (patternFiles.length === 0) return [];
    return this.generateOutput(patternFiles, outputDir, engine);
  }

  private generateOutput(
    patternFiles: readonly string[],
    outputDir: string,
    engine: TemplateEngine,
  ): string[] {
    const rendered = this.renderContents(patternFiles, engine);
    const refsDir = this.buildRefsDir(outputDir);
    const results = this.flushPatterns(patternFiles, rendered, refsDir);
    const consolidated = this.buildConsolidatedPath(outputDir);
    results.push(this.flushConsolidated(rendered, consolidated));
    return results;
  }

  private renderContents(
    patternFiles: readonly string[],
    engine: TemplateEngine,
  ): string[] {
    return patternFiles.map((srcFile) => {
      const content = fs.readFileSync(srcFile, "utf-8");
      return engine.replacePlaceholders(content);
    });
  }

  private buildRefsDir(outputDir: string): string {
    return path.join(
      outputDir, SKILLS_DIR, PATTERNS_SKILL_DIR, REFERENCES_DIR,
    );
  }

  private buildConsolidatedPath(outputDir: string): string {
    return path.join(
      outputDir, SKILLS_DIR, PATTERNS_SKILL_DIR, CONSOLIDATED_FILENAME,
    );
  }

  private flushPatterns(
    patternFiles: readonly string[],
    rendered: readonly string[],
    destDir: string,
  ): string[] {
    const results: string[] = [];
    for (let i = 0; i < patternFiles.length; i++) {
      const srcFile = patternFiles[i]!;
      const content = rendered[i]!;
      const category = path.basename(path.dirname(srcFile));
      const targetDir = path.join(destDir, category);
      fs.mkdirSync(targetDir, { recursive: true });
      const destFile = path.join(targetDir, path.basename(srcFile));
      fs.writeFileSync(destFile, content, "utf-8");
      results.push(destFile);
    }
    return results;
  }

  private flushConsolidated(
    rendered: readonly string[],
    destPath: string,
  ): string {
    fs.mkdirSync(path.dirname(destPath), { recursive: true });
    const merged = rendered.join(SECTION_SEPARATOR);
    fs.writeFileSync(destPath, merged, "utf-8");
    return destPath;
  }
}
